runtime/compiler/aarch64/runtime/PicBuilder.spp

```diff
@@ -243,8 +251,70 @@ _interpreterUnresolvedInstanceDataGlue:
 _interpreterUnresolvedInstanceDataStoreGlue:
 	hlt	#0	// Not implemented yet
 
+// For vitual unresolved call, we generate following instructions
+//  movz  x9, #0
+//  movk  x9, #0, LSL #16
+//  movk  x9, #0, LSL #32
+//  movk  x9, #0, LSL #48
+//  ldr   x9, [vftReg, #0]
+//  b     VirtualUnresolvedSnippet ; change this to "blr x9"
+//
+// If resolved index is in 12bit range, we encode the immediate value in the ldr instruction.
+// If resolved index is wider, we encode the immediate value to movz and movk instructions and change the ldr instruction to "ldr x9, [vftReg, x9]".
+// Then, we change the b instruction to "blr x9"
+//
 _virtualUnresolvedHelper:
+	stp	x0, x1, [J9SP, #-64]!				// save parameter regs
+	stp	x2, x3, [J9SP, #16]
+	stp	x4, x5, [J9SP, #32]
+	stp	x6, x7, [J9SP, #48]
+	add	x0, x30, #J9TR_UVCSnippet_CP			// get CP/index pair pointer
+	ldr	x1, [x30, #J9TR_UVCSnippet_returnAddress]	// get code cache RA
+	mov	x4, x1						// protect LR in x4 (in L_commonLookupException, it is expected)
+	mov	x2, x1						// preserve code cache RA
+	bl	jitResolveVirtualMethod				// resolve the method, return value = vTable offset
+	cbz	x0, L_commonLookupException			// if resolve failed, throw the exception
+	neg	x1, x0						// convert vTable offset to positive value
+	cmp	x1, #4095					// test if index is within the 12bit range
+	bgt	L_virtualUnresolvedHelperWideIndex		// If index > 4095
+	sub	x0, x2, #8					// get address of the ldr instruction
```

Comment:
It looks like the following code is patching up an instruction sequence in the mainline code (i.e., not the virtual call snippet).  Since that code is not hooked up yet, can you describe what that instruction sequence looks like?  It will help understand what's going on here.  Thanks.

Reply:
Please take a look at the comment lines above the label `_virtualUnresolvedHelper:` for the mainline code that is patched.